modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java

```diff
@@ -723,6 +725,25 @@ private void initCachesOnLocalJoin() throws IgniteCheckedException {
         }
 
         cctx.cache().startCachesOnLocalJoin(caches, initialVersion());
+
+        ensureClientCachesStarted();
+    }
+
+    /**
+     * Start client caches if absent.
+     */
+    private void ensureClientCachesStarted() {
+        GridCacheProcessor cacheProcessor = cctx.cache();
+        List<CacheConfiguration> notStartedCacheConfigs = new ArrayList<>();
+        for (CacheConfiguration cCfg : cctx.gridConfig().getCacheConfiguration()) {
+            if (!cacheProcessor.cacheNames().contains(cCfg.getName())) {
```

Comment:
I would recommend extracting cacheProcessor.cacheNames() to a variable and converting it to a Set in process to speed up lookups. Because in cases when there's hundred of caches there will be a performance issue.